File: server/apps/api/src/animedownloader_api/routes/download_jobs.py
from typing import Annotated
from uuid import UUID

from animedownloader_download import DownloadJob, DownloadJobService, DownloadJobStatus
from animedownloader_qbittorrent import QBittorrentClient
from fastapi import APIRouter, Depends, Query

from animedownloader_api.dependencies import (
    get_download_job_service,
    get_qbittorrent_client,
)
from animedownloader_api.download_control import DownloadControlService
from animedownloader_api.schemas import (
    DownloadJobListItemResponse,
    DownloadJobListResponse,
    DownloadJobResponse,
)

router = APIRouter(prefix="/api/download-jobs", tags=["download-jobs"])
QBittorrentClientDependency = Annotated[
    QBittorrentClient,
    Depends(get_qbittorrent_client),
]
DownloadJobServiceDependency = Annotated[
    DownloadJobService,
    Depends(get_download_job_service),
]


@router.get("", response_model=DownloadJobListResponse)
async def list_download_jobs(
    service: DownloadJobServiceDependency,
    statuses: Annotated[
        list[DownloadJobStatus] | None,
        Query(alias="status"),
    ] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 50,
) -> DownloadJobListResponse:
    result = await service.list_jobs(
        statuses=tuple(statuses) if statuses else None,
        page=page,
        page_size=page_size,
    )
    return DownloadJobListResponse(
        items=[
            DownloadJobListItemResponse.model_validate(
                {
                    **DownloadJobResponse.model_validate(item.job).model_dump(),
                    "anime_id": item.anime.id,
                    "anime_title": item.anime.title,
                    "episode_number": item.episode.episode_number,
                    "episode_title": item.episode.title,
                }
            )
            for item in result.items
        ],
        page=page,
        page_size=page_size,
        total=result.total,
        has_more=page * page_size < result.total,
    )


@router.get("/{job_id}", response_model=DownloadJobResponse)
async def get_download_job(
    job_id: UUID,
    service: DownloadJobServiceDependency,
) -> DownloadJob:
    return await service.get_job(job_id)


@router.post("/{job_id}/pause", response_model=DownloadJobResponse)
async def pause_download_job(
    job_id: UUID,
    service: DownloadJobServiceDependency,
    torrent_client: QBittorrentClientDependency,
) -> DownloadJob:
    return await DownloadControlService(service, torrent_client).pause(job_id)


@router.post("/{job_id}/resume", response_model=DownloadJobResponse)
async def resume_download_job(
    job_id: UUID,
    service: DownloadJobServiceDependency,
    torrent_client: QBittorrentClientDependency,
) -> DownloadJob:
    return await DownloadControlService(service, torrent_client).resume(job_id)


@router.post("/{job_id}/cancel", response_model=DownloadJobResponse)
async def cancel_download_job(
    job_id: UUID,
    service: DownloadJobServiceDependency,
    torrent_client: QBittorrentClientDependency,
) -> DownloadJob:
    return await DownloadControlService(service, torrent_client).cancel(job_id)


@router.delete("/{job_id}", status_code=204)
async def delete_download_job(
    job_id: UUID,
    service: DownloadJobServiceDependency,
    torrent_client: QBittorrentClientDependency,
) -> None:
    await DownloadControlService(service, torrent_client).delete(job_id)

File: web/apps/web/src/entities/download/api/downloadJobs.ts
import { z } from "zod";
import { deleteJson, getJson, postJson } from "../../../shared/api/client";
import type {
  DownloadJob,
  DownloadJobListParams,
  DownloadJobListResponse,
} from "../model/types";

const downloadJobSchema = z.object({
  id: z.uuid(),
  episode_id: z.uuid(),
  status: z.enum([
    "pending",
    "downloading",
    "paused",
    "completed",
    "failed",
    "cancelled",
  ]),
  downloaded_bytes: z.number().int().nonnegative(),
  total_bytes: z.number().int().nonnegative().nullable(),
  attempt_count: z.number().int().nonnegative(),
  error_message: z.string().nullable(),
  started_at: z.coerce.date().nullable(),
  completed_at: z.coerce.date().nullable(),
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
});

const downloadJobListItemSchema = downloadJobSchema.extend({
  anime_id: z.uuid(),
  anime_title: z.string(),
  episode_number: z.number().int().positive(),
  episode_title: z.string(),
});

const downloadJobListResponseSchema = z.object({
  items: z.array(downloadJobListItemSchema),
  page: z.number().int().positive(),
  page_size: z.number().int().positive(),
  total: z.number().int().nonnegative(),
  has_more: z.boolean(),
});

export class DownloadJobResponseError extends Error {
  constructor(readonly issues: z.core.$ZodIssue[]) {
    super(
      "Invalid download job API response: " +
        issues
          .map(
            (issue) =>
              (issue.path.join(".") || "<root>") + ": " + issue.message,
          )
          .join("; "),
    );
    this.name = "DownloadJobResponseError";
  }
}


export async function listDownloadJobs(
  params: DownloadJobListParams = {},
): Promise<DownloadJobListResponse> {
  const search = new URLSearchParams();
  for (const status of params.statuses ?? []) {
    search.append("status", status);
  }
  search.set("page", String(params.page ?? 1));
  search.set("page_size", String(params.pageSize ?? 100));

  const result = downloadJobListResponseSchema.safeParse(
    await getJson("/api/download-jobs?" + search.toString(), {
      signal: params.signal,
    }),
  );
  if (!result.success) {
    throw new DownloadJobResponseError(result.error.issues);
  }
  return result.data;
}

export async function getLatestEpisodeDownloadJob(
  episodeId: string,
  signal?: AbortSignal,
): Promise<DownloadJob | null> {
  const payload = await getJson(
    "/api/episodes/" + episodeId + "/download-jobs/latest",
    { signal },
  );

  if (payload === null) {
    return null;
  }

  return parseDownloadJob(payload);
}

export async function createEpisodeDownloadJob(
  episodeId: string,
  signal?: AbortSignal,
): Promise<DownloadJob> {
  return parseDownloadJob(
    await postJson(
      "/api/episodes/" + episodeId + "/download-jobs",
      {},
      { signal },
    ),
  );
}

function parseDownloadJob(payload: unknown): DownloadJob {
  const result = downloadJobSchema.safeParse(payload);
  if (!result.success) {
    throw new DownloadJobResponseError(result.error.issues);
  }
  return result.data;
}


export async function pauseDownloadJob(
  jobId: string,
  signal?: AbortSignal,
): Promise<DownloadJob> {
  return parseDownloadJob(
    await postJson("/api/download-jobs/" + jobId + "/pause", {}, { signal }),
  );
}

export async function resumeDownloadJob(
  jobId: string,
  signal?: AbortSignal,
): Promise<DownloadJob> {
  return parseDownloadJob(
    await postJson("/api/download-jobs/" + jobId + "/resume", {}, { signal }),
  );
}

export async function cancelDownloadJob(
  jobId: string,
  signal?: AbortSignal,
): Promise<DownloadJob> {
  return parseDownloadJob(
    await postJson("/api/download-jobs/" + jobId + "/cancel", {}, { signal }),
  );
}

export async function deleteDownloadJob(
  jobId: string,
  signal?: AbortSignal,
): Promise<void> {
  await deleteJson("/api/download-jobs/" + jobId, { signal });
}
